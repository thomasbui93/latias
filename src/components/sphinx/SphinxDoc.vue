<template>
  <div class="sphinx-doc">
    <div class="sphinx-doc__title">{{question}}</div>
    <div class="sphinx-doc__content">
      <div class="sphinx-doc__keywords">
        <div class="sphinx-tag"
        v-for="(keyword, index) in keywords" :key="index" @click="toggleKeyword(keyword)">
          {{keyword}}
        </div>
      </div>
      <div class="sphinx-doc__category">
        <strong>Category: </strong><span>{{categories}}</span>
      </div>
      <div class="sphinx-doc__level">
        <strong>Level: </strong><span class="is-lowercase" @click="toggleLevel">{{level}}</span>
      </div>
      <div class="sphinx-doc__description">
        <strong>Answer: </strong>{{answer}}
      </div>
      <div class="sphinx-doc__meta">
        Added: {{ createdAt | dateNormalized }}
      </div>
    </div>
  </div>
</template>

<style lang="scss">
.sphinx-doc {
  box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
  transition: all 0.3s cubic-bezier(.25,.8,.25,1);

  &__title {
    position: relative;
    width: 100%;
    border-bottom: 1px solid $grey-light;
    padding: 0.75rem;
    font-weight: bold;
  }
  &__content {
    padding: 0.75rem;
  }
}
</style>

<script>
export default {
  props: {
    question: {
      type: String,
      default: '',
    },
    keywords: {
      type: Array,
      default: () => [],
    },
    answer: {
      type: String,
      default: '',
    },
    level: {
      type: String,
      default: '',
    },
    createdAt: {
      type: String,
      default: '',
    },
    categories: {
      type: String,
      default: '',
    },
  },
  methods: {
    toggleKeyword(keyword) {
      this.$emit('toggleKeyword', keyword);
    },
    toggleLevel() {
      this.$emit('toggleLevel', this.level);
    },
  },
};
</script>
