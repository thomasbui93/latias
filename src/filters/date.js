import Vue from 'vue';
import dayjs from 'dayjs';

export const normalizedDate = (value) => {
  if (!value) return '';
  const date = dayjs(value);
  return date.isValid() ? dayjs(value).format('YYYY-MM-DD') : '';
};

export default Vue.filter('dateNormalized', normalizedDate);
