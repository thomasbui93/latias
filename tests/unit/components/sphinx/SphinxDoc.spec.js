import SphinxDoc from '@/components/sphinx/SphinxDoc.vue';
import { mount, shallowMount } from '@vue/test-utils';
import '@/filters/date';

describe('SphinxDoc component', () => {
  it('match snapshot', () => {
    const wrapper = shallowMount(SphinxDoc, {
      propsData: {
        question: 'test question',
        keywords: ['keyword1', 'keyword2'],
        categories: 'category',
        answer: 'answer',
        level: 'BEGINER',
        createdAt: new Date().toString(),
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('render empty tags if keywords are missing', () => {
    const wrapper = shallowMount(SphinxDoc, {
      propsData: {
        question: 'test question',
        categories: 'category',
        answer: 'answer',
        level: 'BEGINER',
        createdAt: new Date().toString(),
      },
    });
    expect(wrapper.vm.keywords.length).toBe(0);
  });

  it('click on level to emit toggleLevel', () => {
    const wrapper = mount(SphinxDoc, {
      propsData: {
        question: 'test question',
        keywords: ['keyword1', 'keyword2'],
        categories: 'category',
        answer: 'answer',
        level: 'BEGINER',
        createdAt: new Date().toString(),
      },
    });
    wrapper.find('.sphinx-doc__level .is-lowercase').trigger('click');
    expect(wrapper.emitted().toggleLevel[0]).toEqual(['BEGINER']);
  });

  it('click on tag to emit toggleKeyword', () => {
    const wrapper = mount(SphinxDoc, {
      propsData: {
        question: 'test question',
        keywords: ['keyword1', 'keyword2'],
        categories: 'category',
        answer: 'answer',
        level: 'BEGINER',
        createdAt: new Date().toString(),
      },
    });
    wrapper.find('.sphinx-tag:nth-child(2)').trigger('click');
    expect(wrapper.emitted().toggleKeyword[0]).toEqual(['keyword2']);
  });
});
