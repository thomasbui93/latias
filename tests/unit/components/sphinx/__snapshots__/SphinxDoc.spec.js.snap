// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SphinxDoc component match snapshot 1`] = `
<div class="sphinx-doc">
  <div class="sphinx-doc__title">test question</div>
  <div class="sphinx-doc__content">
    <div class="sphinx-doc__keywords">
      <div class="sphinx-tag">keyword1</div>
      <div class="sphinx-tag">keyword2</div>
    </div>
    <div class="sphinx-doc__category"><strong>Category:</strong> <span>category</span></div>
    <div class="sphinx-doc__level"><strong>Level:</strong> <span class="is-lowercase">BEGINER</span></div>
    <div class="sphinx-doc__description"><strong>Answer:</strong>
      answer
    </div>
    <div class="sphinx-doc__meta">Added: 2019-01-09</div>
  </div>
</div>
`;
